CodeGeneration/Sources/generate-swiftsyntax/templates/swiftsyntax/SyntaxNodeFile.swift

```diff
@@ -0,0 +1,432 @@
+              expression: DictionaryExprSyntax {
+                for child in node.children {
+                  if child.isOptional {
+                    DictionaryElementSyntax(
+                      leadingTrivia: .newline,
+                      keyExpression: ExprSyntax(#""\#(raw: child.swiftName)""#),
+                      valueExpression: ExprSyntax("\(raw: child.swiftName).map(Syntax.init)?.asProtocol(SyntaxProtocol.self) as Any"))
+                  } else {
+                    DictionaryElementSyntax(
+                      leadingTrivia: .newline,
+                      keyExpression: ExprSyntax(#""\#(raw: child.swiftName)""#),
+                      valueExpression: ExprSyntax("Syntax(\(raw: child.swiftName)).asProtocol(SyntaxProtocol.self)"))
+                  }
+                }
+              }
+            )
+          })
+        }
+      }
+    }
+  }
+}
+
+private func generateSyntaxChildChoices(for child: Child) throws -> EnumDeclSyntax {
+  return try EnumDeclSyntax("public enum \(raw: child.name): SyntaxChildChoices") {
+    for choice in child.nodeChoices {
+      DeclSyntax("case `\(raw: choice.swiftName)`(\(raw: choice.typeName))")
+    }
+
+    try! VariableDeclSyntax("public var _syntaxNode: Syntax") {
+      try! SwitchStmtSyntax("switch self") {
+        for choice in child.nodeChoices {
+          SwitchCaseSyntax("case .\(raw: choice.swiftName)(let node):") {
+            StmtSyntax("return node._syntaxNode")
+          }
+        }
+      }
+    }
+
+    DeclSyntax("init(_ data: SyntaxData) { self.init(Syntax(data))! }")
+
+    for choice in child.nodeChoices {
+      if let choiceNode = SYNTAX_NODE_MAP[choice.syntaxKind], choiceNode.isBase {
+        DeclSyntax("""
+          public init<Node: \(raw: choiceNode.name)Protocol>(_ node: Node) {
+            self = .\(raw: choice.swiftName)(\(raw: choiceNode.name)(node))
+          }
+          """)
+
+      } else {
+        DeclSyntax("""
+          public init(_ node: \(raw: choice.typeName)) {
+            self = .\(raw: choice.swiftName)(node)
+          }
+          """)
+      }
+    }
+
+    try! InitializerDeclSyntax("public init?<S: SyntaxProtocol>(_ node: S)") {
+      for choice in child.nodeChoices {
+        StmtSyntax("""
+          if let node = node.as(\(raw: choice.typeName).self) {
+            self = .\(raw: choice.swiftName)(node)
+            return
+          }
+          """)
+      }
+
+      StmtSyntax("return nil")
+    }
+
+    try! VariableDeclSyntax("public static var structure: SyntaxNodeStructure") {
+      let choices = ArrayExprSyntax {
+        for choice in child.nodeChoices {
+          ArrayElementSyntax(
+            expression: ExprSyntax(".node(\(raw: choice.typeName).self)"))
+        }
+      }
+
+      StmtSyntax("return .choices(\(choices))")
+    }
+  }
+}
+
+private func generateInitializerDecl(children: [Child], optionalBaseAsMissing: Bool, currentIndentation: String = "") -> DeclSyntax {
```

Comment:
I would expect that BasicFormat handles indentation so we shouldn’t need `currentIndentation` and all the indentation handling below.

---

Also, I would rename this to `generateInitializerDeclHeader` because it’s not generating the full initializer decl, just its header without the body. This could also return a `PartialSyntaxNodeString` instead of a `DeclSyntax` to indicate that the returned value doesn’t have a body.